### PR TITLE
Parse dates better in sqlcmd.

### PR DESCRIPTION
**Pull request by BillWhite**
The documented forms for VoltDB dates are 'yyyy-mm-dd HH:MM.sss',
'yyyy-MM-dd HH:MM' and 'yyyy-MM-dd'.  The previous pattern did not
work for the latter two.  This commit fixes this by trying all three
patterns.  There seems to be no way to specify that a part of the
pattern is optional.

I changed the parsing strategy, and just use the one for TimestampType.  This
gives us microsecond granularity, which is what we seem to be advertising,
and it parses all the formats we advertise.

In this commit I also added a fix to ENG-10369.  It was really obvious,
and easier to fix it than to describe how to fix it in the ticket.

https://issues.voltdb.com/browse/ENG-10031

### CODE REVIEW

File: tests/sqlcmd/baselines/simple/sqlcmd1.outbaseline [hunk 1]
```diff
@@ -25,3 +25,50 @@ C
...
+insert into FOO values ('2015-11-11 00:00:00');
+(Returned 1 rows in #.##s)
+
+insert into FOO values ('2015-11-11');
+(Returned 1 rows in #.##s)
+
+exec FOO.insert '2015-11-11 00:00:00.000000';
+(Returned 1 rows in #.##s)
+
+exec FOO.insert '2015-11-11 00:00:00.0';
```
Comment (paulmartel) [inline tests/sqlcmd/baselines/simple/sqlcmd1.outbaseline]: A stronger test would use non-zero values and verify that they come back intact.

File: src/frontend/org/voltdb/parser/SQLParser.java [hunk 2]
```diff
@@ -540,7 +540,9 @@ private Exception(Throwable cause, String message, Object... args)
             "\\s*",              // extra spaces
             Pattern.MULTILINE + Pattern.CASE_INSENSITIVE);
 
-    private static final SimpleDateFormat DateParser = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
+    private static final SimpleDateFormat FullDateParser = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
```
Comment (paulmartel) [inline src/frontend/org/voltdb/parser/SQLParser.java]: We should have tests ("negative tests" if necessary) for a trailing decimal point after the seconds, and for one, two, three, or four digits (not all 0s) after the decimal point.

File: src/frontend/org/voltdb/parser/SQLParser.java [hunk 3]
```diff
@@ -540,7 +540,9 @@ private Exception(Throwable cause, String message, Object... args)
...
+    private static final SimpleDateFormat WholeSecondDateParser = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
Comment (paulmartel) [inline src/frontend/org/voltdb/parser/SQLParser.java]: Also test for 1-digit month, day, and hour?
Test (negative?) for 1 digit second?

File: src/frontend/org/voltdb/parser/SQLParser.java [hunk 4]
```diff
@@ -540,7 +540,9 @@ private Exception(Throwable cause, String message, Object... args)
...
+    private static final SimpleDateFormat DayDateParser = new SimpleDateFormat("yyyy-MM-dd");
```
Comment (paulmartel) [inline src/frontend/org/voltdb/parser/SQLParser.java]: Test for 1-digit month, day?

**Comment by paulmartel** [comment]
End of review round.

File: tests/sqlcmd/scripts/simple/parsedates.in [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+-- in sqlcmdtest, because they cause sqlcmd to fail.
+insert into FOO values ('20150-11-11 10:10:10');
+insert into FOO values ('201-11-11 10:10:10');
+insert into FOO values ('2015-111-11 10:10:10');
+insert into FOO values ('2015-1-11 10:10:10');
+insert into FOO values ('2015-11-111 10:10:10');
+insert into FOO values ('2015-11-1 10:10:10');
+
+-- Now try the same things as before, but with a CRUD procedure.
+-- This is problematic because sqlcmd tries to parse the datetime.
```
Comment (paulmartel) [inline tests/sqlcmd/scripts/simple/parsedates.in]: Should this be "was problematic"? Shouldn't the negative cases also be repeated for this code path?

File: tests/sqlcmd/scripts/simple/parsedates.in [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+insert into FOO values ('2015-11-11 10:11:12.12345');
+insert into FOO values ('2015-11-11 10:11:12.1234');
+insert into FOO values ('2015-11-11 10:11:12.123');
+insert into FOO values ('2015-11-11 10:11:12.12');
+insert into FOO values ('2015-11-11 10:11:12.1');
+insert into FOO values ('2015-11-11 10:11:12');
+insert into FOO values ('2015-11-11');
+
+-- Try some negative cases.  We can't actually do these
+-- in sqlcmdtest, because they cause sqlcmd to fail.
```
Comment (paulmartel) [inline tests/sqlcmd/scripts/simple/parsedates.in]: These are negative test cases -- that part is clear.
The remainder of the comment is confusing. It sounds like it is paradoxically saying "this test does not work".
A more reasonable interpretation seems to make it redundant with the identification of these statements as negative test cases -- like it's trying to define what a negative test case is.

**Comment by paulmartel** [comment]
End of review round.

File: tests/sqlcmd/scripts/simple/parsedates.in (hunk 5 above)
Comment (BillWhite) [inline tests/sqlcmd/scripts/simple/parsedates.in]: I fixed the comment and added negative cases.

File: tests/sqlcmd/scripts/simple/parsedates.in (hunk 6 above)
Comment (BillWhite) [inline tests/sqlcmd/scripts/simple/parsedates.in]: I fixed the comment and added some new negative tests.

File: src/frontend/org/voltdb/parser/SQLParser.java (hunk 4 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/parser/SQLParser.java]: 1 digit days, months, hours, minutes and seconds are apparently legal with the TimestampType parser, which is from JDBC.

File: src/frontend/org/voltdb/parser/SQLParser.java (hunk 3 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/parser/SQLParser.java]: 1 digit days, months, hours, minutes and seconds are apparently legal with the TimestampType parser, which is from JDBC.

**Comment by BillWhite** [comment]
I'm squash merging this separately.

**Pull request closed without merging**